Fixed #198 Created Attendee list API

Apalakimmortal opened the pull request:
# Closes #198 
<!--- Provide an overall summary of the pull request -->

### Changes
<!--- More detailed and granular description of changes -->
<!--- These should likely be gathered from commit message summaries -->
- <ONE>
- <TWO>

### Type of change
- [ ] Bug fix
- [x] Feature update
- [ ] Breaking change
- [ ] Documentation update

### Flags
<!--- Provide context or concerns a reviewer should be aware of -->
- <ONE>
- <TWO>

### Demo
<!--- Provide an easily accessible demonstration of the changes, if applicable (preferably in png/gif format) -->
- <ONE>
- <TWO>

### How has this been tested?
<!---[Describe the tests that you ran to verify your changes]-->
- <ONE>
- <TWO>

### Author Checklist
- [ ] Code has been commented, particularly in hard-to-understand areas 
- [x] Changes generate no new warnings
- [ ] Vital changes have been captured in unit and/or integration tests
- [ ] New and existing unit tests pass locally with my changes
- [ ] Documentation has been extended, if necessary
- [ ] Merging to `main` from `fork:branchname`

### Additional context 
<!--- [Add any other context about the PR here] -->
- <ONE>
- <TWO>

salty-ivy commented:
@Apalakimmortal  is this ready for review?? also can you change the description with context section providing an abstract context about the changes you have made.

Apalakimmortal commented:
Yes bhaiya it is ready for reviewed. I have made changes in EventRegistration Model in which I have added 2 fields 
1. first_time_attende
2.attende_type
First time attende is when a user register for a event 1st time and attende type in which there are 4 options 
Guest,Host,Speaker and Volunteer

Along with Model i have also changed the Serealizer file and admin.py file for registering 
Then i created views in which i am sending API which consist of 1. Name of Event 2. Total Number of attende 3. First time attende count and info about all the people who are attending that particular event.

salty-ivy commented on backend/djangoindia/api/views/event.py:
```diff
@@ -1,18 +1,46 @@
+
+        response_data = {
+            'event_name': event.name,
+            'total_attendees': queryset.count(),
```
calling `Count()` twice is inefficient and increases DB hits use `aggregate` to fetch all sorts of kind in a single call.

```python
from django.db.models import Q 

attendee_data = queryset.aggregate(
            total_attendees=Count('id'),
            first_time_attendees=Count('id', filter=Q(first_time_attendee=True))
        )

        response_data = {
            'event_name': event.name,
            'total_attendees': attendee_data['total_attendees'],
            'first_time_attendees': attendee_data['first_time_attendees'],
            'attendees': serializer.data
        }
```

salty-ivy commented on backend/djangoindia/api/serializers/event.py:
```diff
@@ -55,6 +55,23 @@ class EventRegistrationSerializer(serializers.Serializer):
+    first_time_attendee = serializers.BooleanField(default= True)
+    attendee_type = serializers.ChoiceField(choices=EventRegistration.AttendeeType)
+    
+    class EventAttendeeSerializer(serializers.ModelSerializer):
```
why do we need an inner class `ModelSerializer`  that doesn't connect to any model?

Apalakimmortal replied:
<img width="211" alt="khi" src="https://github.com/user-attachments/assets/a209898e-a4d7-4d11-b958-6e7573afe441">
Bhaiya I am not able to change it to different attende_type from admin panel. Can u please help me with that

Apalakimmortal replied:
@salty-ivy

salty-ivy reviewed (changes requested):
looks solid, added some comments and changes, make sure to test them out manually before publishing.

CC: @DevilsAutumn

Apalakimmortal commented:
Bhaiya do i need to change the things u suggested.

Apalakimmortal commented:
Will do it by tomorrow

DevilsAutumn commented on backend/djangoindia/api/urls/event.py:
```diff
@@ -1,8 +1,8 @@
     path("events/<slug:slug>/", EventAPIView.as_view({"get":"retrieve","post":"post"}), name="get_event"),
-]
\ No newline at end of file
+    path("events/<uuid:event_id>/attendees/", EventAttendeeViewSet.as_view({"get": "list"}), name="get_event_attendee"),  
```
use slug instead of event is

DevilsAutumn commented on backend/djangoindia/api/views/event.py:
```diff
@@ -1,18 +1,50 @@
+        event_id = self.kwargs.get('event_id')
+        er =  EventRegistration.objects.filter(
+            event_id=event_id,
+            # include_in_attendee_list=True
```
comments do not get executed .

DevilsAutumn commented on backend/djangoindia/settings/base.py:
```diff
@@ -21,7 +21,7 @@
 BASE_DIR = Path(__file__).resolve().parent.parent.parent
 
-DOCKERIZED = int(os.environ.get("DOCKERIZED", 0)) == 1
+DOCKERIZED = (os.environ.get("DOCKERIZED", 0)) == 1
```
Why?

DevilsAutumn commented on backend/djangoindia/api/serializers/event.py:
```diff
@@ -55,6 +55,23 @@ class EventRegistrationSerializer(serializers.Serializer):
     other_links = serializers.URLField(required=False, allow_blank=True)
     rsvp = serializers.BooleanField(default=False)
     include_in_attendee_list = serializers.BooleanField(default=False)
+    first_time_attendee = serializers.BooleanField(default= True)
+    attendee_type = serializers.ChoiceField(choices=EventRegistration.AttendeeType)
+    
+    class EventAttendeeSerializer(serializers.Serializer):
+        full_name = serializers.SerializerMethodField()
+    
+    class Meta:
+        model = EventRegistration
+        fields = [
+            'full_name', 'professional_status', 'organization',
+            'linkedin', 'github', 'twitter', 'first_time_attendee',
+            'attendee_type'
+        ]
 
     def create(self, validated_data):
         return EventRegistration.objects.create(**validated_data)
+    
+    def get_full_name(self, obj):
+        return f"{obj.first_name} {obj.last_name}"
+
```
I think you have mixed 2 serializers here? `EventRegistrationSerializer` and `EventAttendeeSerializer` should be different . `EventRegistrationSerializer` is being used only to validate the event registrations data.

DevilsAutumn reviewed (changes requested):
Looks good! Left few comments.

DevilsAutumn commented on backend/djangoindia/api/views/event.py:
```diff
@@ -1,19 +1,55 @@
+        event_slug = self.kwargs.get('event_slug')
+        event = get_object_or_404(Event, slug=event_slug)
+        queryset = EventRegistration.objects.filter(
+            event__slug=event_slug
```
we need to add `include_in_attendee_list=True` in this.

DevilsAutumn commented on backend/djangoindia/api/views/event.py:
```diff
@@ -1,19 +1,55 @@
 
 # Create your views here.
+class EventAttendeeViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
+    serializer_class = EventRegistrationSerializer 
```
```suggestion
    serializer_class = EventAttendeeSerializer 

```

DevilsAutumn commented on backend/djangoindia/api/serializers/event.py:
```diff
@@ -55,6 +55,21 @@ class EventRegistrationSerializer(serializers.Serializer):
+class EventAttendeeSerializer(serializers.Serializer):
+    full_name = serializers.SerializerMethodField()
+
+    class Meta:
+        model = EventRegistration
+        fields = [
+            'full_name', 'professional_status', 'organization',
+            'linkedin', 'github', 'twitter', 'first_time_attendee',
+            'attendee_type',
+    ]
```
This is kindof correct but we specify field type of each field explicitly as you can see in other serializers due to performance issues in ModelSerializer.

DevilsAutumn commented on backend/djangoindia/api/serializers/event.py:
```diff
@@ -55,6 +55,21 @@ class EventRegistrationSerializer(serializers.Serializer):
     other_links = serializers.URLField(required=False, allow_blank=True)
     rsvp = serializers.BooleanField(default=False)
     include_in_attendee_list = serializers.BooleanField(default=False)
+    first_time_attendee = serializers.BooleanField(default= True)
+    attendee_type = serializers.ChoiceField(choices=EventRegistration.AttendeeType)
```
we can remove this as it will be controlled from admin panel and model class.
```suggestion

```

DevilsAutumn reviewed (changes requested):
Few more comments. Also you can delete all migrations and create one single migration for all the changes in this PR.

DevilsAutumn commented:
@Apalakimmortal Thanks 👍  I've rebase this PR to main and pushed few changes.

salty-ivy reviewed (approved):
LGTM. Great work @DevilsAutumn @Apalakimmortal

DevilsAutumn commented:
@salty-ivy Thankyou for reviewing 👍
